dali/operators/reader/numpy_reader_op.cc

```diff
@@ -226,6 +234,111 @@ DALI_SCHEMA(NumpyReader)
+                      DomainTimeRange::kRed);
+  DataReader<CPUBackend, NumpyFileWrapper>::Prefetch();
+
+  if (!dont_use_mmap_)
+    return;
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+
+  string previous_path;
+  for (unsigned idx = 0; idx < curr_batch.size(); ++idx) {
+    auto &target = curr_batch[idx];
+    if (target->data.shares_data()) {
+      target->data.Reset();
+    }
+    // if we pad last batch but we duplicate the samples from the previous one - a case
+    // with multiple unequal shards where we need to create a full duplicated batch
+    // so we need to reopen the file and seek
+    if (!target->current_file) {
+      target->current_file = FileStream::Open(target->filename, false, false, use_o_direct_);
+      target->current_file->SeekRead(target->data_offset);
+    }
+    if (use_o_direct_) {
+      /*
+       * the offset to the data we are interested in
+       *                  |<-                       aligned_len                    ->|
+       *                  |                                                          |
+       *                  |                               |<- target->nbytes ->|     |
+       *                  |<-    target_data_offset     ->|                    |     |
+       *     |<-          target->data_offset           ->|
+       *     -------------********************************XXXXXXXXXXXXXXXXXXXXXX******
+       *     |            |                               |                          |
+       * file_start   block_start/allocated_memory     shared_ptr                   block_end
+       */
+
+      // align the size of data to read accounting for its start
+      auto block_start = align_down(target->data_offset, o_direct_alignm_);
+      auto block_end = align_up(target->data_offset + target->nbytes, o_direct_alignm_);
+      auto aligned_len = align_up(block_end - block_start, o_direct_read_len_alignm_);
+      // offset to the desired data from the block start
+      auto target_data_offset = alignment_offset(target->data_offset, o_direct_alignm_);
+      // allocate the memory that is aligned to the block size, but wrap it into shared ptr using
+      auto resource_mem = mm::alloc_raw_shared<char, mm::memory_kind::host>(aligned_len,
+                                                                            o_direct_alignm_);
+      shared_ptr<void> tmp_mem(
+        resource_mem.get() + target_data_offset,
+        [resource_mem](void *p) {
+          // This is an empty lambda, which is a custom deleter for
+          // std::shared_ptr.
+          // While instantiating shared_ptr, also lambda is instantiated,
+          // making a copy of resource_mem. This way, reference counter of resource_mem
+          // is incremented. Therefore, for the duration of life cycle of
+          // underlying memory in shared_ptr, resource_mem won't free its memory.
+          // It will be freed, when last shared_ptr is deleted.
+        });
```

Comment:
That's what aliasing shared pointers are for.
```suggestion
      shared_ptr<void> tmp_mem(resource_mem, resource_mem.get() + target_data_offset);
```